src/server/mod.rs

```diff
@@ -94,20 +94,19 @@ where B: Body,
 }
 
 /// Error produced by a `Connection`.
-pub enum Error<S>
-where S: MakeService<(), Request<RecvBody>>,
-{
+#[derive(Debug)]
+pub enum Error<E, ME = E> {
```

Comment:
Since `Server` accepts a `MakeService` is there ever a chance that this won't be `S::Error` or `S::MakeError`?

Reply:
> Since `Server` accepts a `MakeService` is there ever a chance that this won't be `S::Error` or `S::MakeError`?

No, and it's a good thing: `Error` does not need to drag the full `MakeService` implementation type into its signature, because it only needs two subdomain error types which are: 1) usually much simpler; 2) are types of actual contents of the `Error` enum, which makes `Debug` and other impls auto-derivable as seen in the commit.

Every meaningful declaration using this `Error` type locks the parameter types into the associated types of `MakeService`. Code that attempts to misuse the type parameters, e.g. by unintentionally swapping them, would either fail type-checking at use of the constructors `Error::NewService` and `Error::Service`, or would look obviously wrong.